engine/classes/Elgg/Database/EntityTable.php

```diff
@@ -217,13 +300,31 @@ public function get($guid, $type = '') {
 			return false;
 		}
 
-		// Check local cache first
-		$new_entity = _elgg_services()->entityCache->get($guid);
-		if ($new_entity) {
+		$guid = (int) $guid;
+
+		$memcache = _elgg_get_memcache('new_entity_cache');
+		
+		$entity = $this->entity_cache->get($guid);
+
+		if (!$entity) {
+			$entity = $memcache->load($guid);
+			// Validate accessibility
+			if ($entity && !elgg_get_ignore_access() && !has_access_to_entity($entity)) {
+				$entity = false;
+			}
+		}
+
+		if (!$entity instanceof ElggEntity) {
+			$entity = false;
+		}
+
+		if ($entity) {
 			if ($type) {
-				return elgg_instanceof($new_entity, $type) ? $new_entity : false;
+				// Verify type of the cached entity
+				return elgg_instanceof($entity, $type) ? $entity : false;
 			}
-			return $new_entity;
+			$this->entity_cache->set($entity);
```

Comment:
Sloppy that even entities pulled from entity_cache are re-saved there, but if it bugs me I can rework later.